src/uberjob/_execution/run_physical.py

```diff
@@ -88,24 +92,24 @@ def prep_run_physical(
         plan, output_node
     )
     plan = prune_source_literals(plan, inplace=inplace)
+    on_started = on_started or identity
+    on_completed = on_completed or identity
+    on_failed = on_failed or _default_on_failed
     retry = retry or identity
 
     def process(node):
         if type(node) is Call:
-            if on_started is not None:
-                on_started(node)
-
+            on_started(node)
+            bound_call = bound_call_lookup[node]
             try:
-                bound_call = bound_call_lookup[node]
                 retry(bound_call.value.run)()
-                bound_call.value = None
             except Exception as e:
-                if on_failed is not None:
-                    on_failed(node, e)
+                on_failed(node, e)
                 raise
+            finally:
+                bound_call.value = None
```

Comment:
Moving this to the finally block fixes a memory leak related to failures. In practice this probably didn't matter.